### DISCUSSION TITLE
Obtain _grad_phi on boundary condition of an Array variable.

### DISCUSSION THREAD

**Discussion by pengfejeac** [idaholab/moose#27324] [Q&A General]
Greetings

I wish to implement $u \cdot \nabla u$ on a boundary for an array variable. 
What I have in mind is I'll put

` _u[_qp] * _grad_u[_qp](component) * _test[_i][_qp]` 

for residual and put

 `_phi[_qp] * _grad_phi[_qp] * _test[_i][_qp]`

for Jacobian, but when I'm trying to obtain the value of `_grad_phi`, with `_grad_phi(_assembly.gradPhiFace(_var)),` I failed. I keep getting an empty object. Did I make a mistake somewhere?

Thanks

**Comment by GiudGiud** [comment]
hello

did you make sure _grad_phi is a reference in your new kernel?
what finite element family are you using?

Guillaume

**Reply by pengfejeac** [reply]
Yep, in the header file, I defined:

`const ArrayVariablePhiGradient & _grad_phi;`

The finite element family I used is first order Lagrange finite element.

**Reply by GiudGiud** [reply]
actually this is a boundary condition right?

I think grad_phi on the face inside the domain would be discontinuous

Can you try to initialize it by retrieving a reference (or pointer) to the array variable and doing this:
```
_grad_phi(_var.gradPhiFace())
```

**Reply by pengfejeac** [reply]
That's right, it's a boundary condition. 

If I do `_grad_phi(_var.gradPhiFace())` and watch `_grad_phi[_qp]` in debugger, it's no longer empty! Now it contains 4 elements and each element is a libMesh vector contains three doubles, with only one of them being nonzero. So I guess three of those four vector represent the gradient vector of the variable, then what would the remaining one be?

**Reply by GiudGiud** [reply]
Are the 4 elements the 4 components of the array variable maybe?

You should Double check the indexing. I don't know that the qp index is the first one. Look for examples in the code

**Reply by pengfejeac** [reply]
The array variable only contains one element,  I used array variable simply because it's easier to work with in terms of my customized kernels. I'll run through my code again and see if there're any missed errors. 

Is there a case/test that involves  u * grad_u on boundary condition that I can refer to? That would be great.

**Reply by GiudGiud** [reply]
There s not u grad_u but there are some examples of using grad_u and indexing into it.

There should be one in the DG kernels

**Reply by pengfejeac** [reply]
Gotcha, thanks a lot!

**Reply by pengfejeac** [reply]
@GiudGiud I spend a great amount of time on this BC but the solver still failed to converge. So I'm thinking about switching to Lagrange Multiplier and I just finished reading the AverageValueConstraint, which brings me another question: What's the difference between using Lagrange multiplier to enforce average value over the domain and actually implement $\int_\Omega u = V$ with residual `(_u[_qp] * _JxW[_qp] * _coord[_qp]  - V) * _test[_qp]` and Jacobian entries of `(_phi[_qp] * _JxW[_qp] * _coord[_qp]  ) * _test[_qp]`?

**Reply by GiudGiud** [reply]
Hello

Before moving on, did you try the troubleshooting instructions there?
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

There are instructions to check the Jacobian and other issues. For example if you are missing a part of the equation on part of the domain, the checks would find singular values, which cause convergence issues. 

> What's the difference between using Lagrange multiplier to enforce average value over the domain and actually implement 
 with residual (_u[_qp] * _JxW[_qp] * _coord[_qp]  - V) * _test[_qp] and Jacobian entries of (_phi[_qp] * _JxW[_qp] * _coord[_qp]  ) * _test[_qp]?

The lagrange multiplier adds another equation to the nonlinear system that gets solved simultaneously. The additional DOF of the Lagrange multiplier is used to write this equation. The average value will be enforced numerically, and should be converged with close to the solver nonlinear tolerance.

Using a boundary condition does not do that, it's adding the coefficients to the existing equations. The average value will only be enforced weakly, and will depend on the mesh discretization I think.

**Reply by pengfejeac** [reply]
> Before moving on, did you try the troubleshooting instructions there?
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

Thanks for pointing out! This is pretty helpful.

> Using a boundary condition does not do that, it's adding the coefficients to the existing equations. The average value will only be enforced weakly, and will depend on the mesh discretization I think.

I implemented the kernel over the whole domain, I should have make that clear, but it's still not working. So I guess weak form of mean value constraint and Lagrange Multiplier are not the same. 

But that brings another question: if I wish to constrain $\int_{\partial \Omega} u \nabla u = V$, where $\partial \Omega$ is the boundary of domain, with Lagrange Multiplier, I will have to implement a customize kernel right? I don't think there's a Lagrange Multiplier that takes a nonlinear constraint over the boundary in Moose.

**Reply by GiudGiud** [reply]
> I will have to implement a customize kernel right? I

I think you'll need to either use the Constraint system to do that or implement a new ScalarKernel

**Reply by pengfejeac** [reply]
Let me dig into Constraint system and see if anything will help! If not, I'll implement a scalar kernel. Thanks!

**Reply by GiudGiud** [reply]
I took a look at the Constraints system and it is more meant to enforce conditions between two parts of the mesh that are touching, whether conformally or not (contact applications). 

I think a ScalarKernel + (maybe changes to a ScalarMultiplierKernel, but probably not) is the way forward.

**Reply by pengfejeac** [reply]
Thanks for the update, I'll take a look at it! I was having trouble logging in Github lately so I can't get back to you on time, sorry about that. 

I also believe Lagrange multiplier is the way to go, it has physical meaning(boundary flux) that would be really useful to me. But under some conditions, I need to define such multiplier on boundary(surface) instead of scalar, is that possible? Actually, as long as I can define a variable on boundary, I can probably figure this out.

**Reply by GiudGiud** [reply]
We cant define variables on boundaries only
You could try to make the logic work out with a variable defined on the lower dimensional block defined on the boundary.
Or define it in the volume, and use trivial kernels for the rest of the volume (like time derivative only)

**Reply by pengfejeac** [reply]
Update! I tried the approach you mentioned but I failed, too many warnings and error to be discussed. So I switch into another approach, which is define a standard variable on one block and array variable on the other block, then use constraints to set them equal to each other on the boundary(interface). That way, I can calculate u * grad_u accurately(hopefully) and I don't have to worry about Lagrange multiplier and define variable on interface. 

It seems like `EqualValueEmbeddedConstraint` is the one to go, so I'm thinking of rewriting it, basically switch the secondary variable into an array variable. If there's something like that been implemented, feel free to let me know 😸
